mozilla_django_oidc/middleware.py

```diff
@@ -0,0 +1,92 @@
+try:
+    from urllib import urlencode
+except ImportError:
+    from urllib.parse import urlencode
+
+import time
+
+import django
+from django.core.urlresolvers import reverse
+from django.http import HttpResponseRedirect
+from django.utils.crypto import get_random_string
+
+from mozilla_django_oidc.utils import absolutify, import_from_settings
+
+
+# Django 1.10 makes changes to how middleware work. In Django 1.10+, we want to
+# use the mixin so that our middleware works as is.
+if django.VERSION >= (1, 10):
+    from django.utils.deprecation import MiddlewareMixin
+else:
+    class MiddlewareMixin(object):
+        pass
+
+
+OIDC_URLS = [
+    reverse(name)
+    for name in [
+        'oidc_authentication_init',
+        'oidc_authentication_callback',
+        'oidc_logout'
+    ]
+]
+
+
+# FIXME(willkg): This doesn't appear to be used anywhere.
+def logout_url():
+    """Log out the user from Auth0."""
+    url = 'https://' + import_from_settings('OIDC_OP_DOMAIN') + '/v2/logout'
+    url += '?' + urlencode({
+        'returnTo': import_from_settings('LOGOUT_REDIRECT_URL', '/'),
+        'client_id': import_from_settings('OIDC_RP_CLIENT_ID')
+    })
+    return url
+
+
+class RefreshIDToken(MiddlewareMixin):
+    """Renews id_tokens after expiry seconds
+
+    For users authenticated with an id_token, we need to check that it's still
+    valid after a specific amount of time and if not, force them to
+    re-authenticate silently.
+
+    """
+    def process_request(self, request):
+        if ((request.method == 'GET' and
+             request.user.is_authenticated() and
+             request.path not in OIDC_URLS and
+             not request.is_ajax())):
+            expiration = request.session.get('oidc_id_token_expiration')
+            if expiration is not None and expiration > time.time():
+                # The id_token is still valid, so we don't have to do anything.
+                return
+
+            # The id_token has expired, so we have to re-authenticate silently.
+            auth_url = import_from_settings('OIDC_OP_AUTHORIZATION_ENDPOINT')
+            client_id = import_from_settings('OIDC_RP_CLIENT_ID')
+            state = get_random_string(import_from_settings('OIDC_STATE_SIZE', 32))
+
+            # Build the parameters as if we were doing a real auth handoff, except
+            # we also include prompt=none.
+            params = {
```

Comment:
nit: Since we are copying the code from the authentication request, should we write a helper function and call it accordingly?